src/NCalc/Expression.cs

```diff
@@ -249,33 +261,32 @@ public Func<TResult> ToLambda<TResult>()
                 body = System.Linq.Expressions.Expression.Convert(body, typeof(TResult));
             }
 
-            var lambda = System.Linq.Expressions.Expression.Lambda<Func<TResult>>(body);
-            return lambda.Compile();
+            return Tuple.Create(body, parameter);
         }
 
-        public Func<TContext, TResult> ToLambda<TContext, TResult>()
+        protected virtual System.Linq.Expressions.Expression ToLinqExpression<TResult>()
         {
-            if (HasErrors())
-            {
-                throw new EvaluationException(Error, ErrorException);
-            }
-
-            if (ParsedExpression == null)
-            {
-                ParsedExpression = Compile(OriginalExpression, (Options & EvaluateOptions.NoCache) == EvaluateOptions.NoCache);
-            }
+            return ToLinqExpressionInternal<Void, TResult>().Item1;
+        }
 
-            var parameter = System.Linq.Expressions.Expression.Parameter(typeof(TContext), "ctx");
-            var visitor = new LambdaExpressionVistor(parameter, Options);
-            ParsedExpression.Accept(visitor);
+        protected virtual Tuple<System.Linq.Expressions.Expression, System.Linq.Expressions.ParameterExpression> ToLinqExpression<TContext, TResult>()
+        {
+            return ToLinqExpressionInternal<TContext, TResult>();
+        }
 
-            var body = visitor.Result;
-            if (body.Type != typeof(TResult))
-            {
-                body = System.Linq.Expressions.Expression.Convert(body, typeof(TResult));
-            }
+        public Func<TResult> ToLambda<TResult>()
+        {
+            System.Linq.Expressions.Expression body = ToLinqExpression<TResult>();
+            var lambda = System.Linq.Expressions.Expression.Lambda<Func<TResult>>(body);
+            return lambda.Compile();
+        }
 
-            var lambda = System.Linq.Expressions.Expression.Lambda<Func<TContext, TResult>>(body, parameter);
+        public Func<TContext, TResult> ToLambda<TContext, TResult>()
+        {
+            var exprAndParamTuple = ToLinqExpression<TContext, TResult>();
+            System.Linq.Expressions.Expression expression = exprAndParamTuple.Item1;
+            System.Linq.Expressions.ParameterExpression parameter = exprAndParamTuple.Item2;

```

Comment:
These independent declarations don't seem to add much clarity if you make the ContextualParameter change I requested.

`var lambda = L.Expression.Lambda<Func<TContext, TResult>>(contextualParameter.Expression, contextualParameter.Context);`